voice/src/main/kotlin/VoiceConnection.kt

```diff
@@ -0,0 +1,146 @@
+import dev.kord.voice.udp.AudioFramePoller
+import dev.kord.voice.udp.AudioFramePollerConfiguration
+import dev.kord.voice.udp.AudioFramePollerConfigurationBuilder
+import kotlinx.coroutines.*
+import mu.KotlinLogging
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
+import kotlin.coroutines.CoroutineContext
+import kotlin.time.TimeSource
+
+/**
+ * Data that represents a [VoiceConnection], these will never change during the life time of a [VoiceConnection].
+ *
+ * @param selfId the id of the bot connecting to a voice channel.
+ * @param guildId the id of the guild that the bot is connecting to.
+ * @param sessionId the id of the current voice session, given by Discord.
+ */
+data class VoiceConnectionData(
+    val selfId: Snowflake,
+    val guildId: Snowflake,
+    val sessionId: String
+)
+
+private val voiceConnectionLogger = KotlinLogging.logger { }
+
+/**
+ * A [VoiceConnection] is an adapter that forms the concept of a voice connection, or a connection between you and Discord voice servers.
+ *
+ * @param gateway the [Gateway] that handles events for the guild this [VoiceConnection] represents.
+ * @param voiceGateway the underlying [VoiceGateway] for this voice connection.
+ * @param data the data representing this [VoiceConnection].
+ * @param voiceGatewayConfiguration the configuration used for [voiceGateway].
+ * @param audioProvider a [AudioProvider] that will provide [AudioFrame] when required.
+ * @param frameInterceptorFactory a factory for [FrameInterceptor]s that is used whenever audio is ready to be sent. See [FrameInterceptor] and [DefaultFrameInterceptor].
+ * @param voiceDispatcher the dispatcher used for this voice connection.
+ */
+@KordVoice
+class VoiceConnection(
+    val gateway: Gateway,
+    val voiceGateway: VoiceGateway,
+    val data: VoiceConnectionData,
+    internal var voiceGatewayConfiguration: VoiceGatewayConfiguration,
+    var audioProvider: AudioProvider,
+    var frameInterceptorFactory: (FrameInterceptorContext) -> FrameInterceptor,
+    voiceDispatcher: CoroutineDispatcher
+) : CoroutineScope {
+    override val coroutineContext: CoroutineContext =
+        SupervisorJob() + voiceDispatcher + CoroutineName("Voice Connection for Guild ${data.guildId.value}")
+
+    private val audioFramePoller = AudioFramePoller(TimeSource.Monotonic, voiceDispatcher)
+
+    init {
+        // handle voice state/server updates (e.g., a move, disconnect, voice server change, etc.)
+        VoiceUpdateEventHandler(this, gateway.events)
+
+        // handle the lifecycle of the udp connection
+        UdpLifeCycleHandler(
+            voiceGateway.events,
+            voiceGateway::send,
+            { audioFramePoller.start(it.withConnection(this)) },
+            voiceDispatcher
+        )
+    }
+
+    /**
+     * Logs into the voice gateway, and begins the process of an audio-ready voice session.
+     */
+    fun connect() {
+        launch {
```

Comment:
I'm not sure if this should be launched, but I believe you should return the reference to the job

Reply:
Kord Core eventually does the same thing in it's logIn method, but ye I can return a reference to the job.